grpc/creds/creds.go

```diff
@@ -51,9 +63,73 @@ func (tc *transportCredentials) ClientHandshake(ctx context.Context, addr string
 	}
 }
 
+func (tc *transportCredentials) peerIsWhitelisted(peerState tls.ConnectionState) error {
+	// If there's no whitelist, all clients are OK
+	if tc.whitelist == nil {
+		return nil
+	}
+
+	// Otherwise its time to start inspecting the peer's `VerifiedChains`
+	chains := peerState.VerifiedChains
+	if len(chains) < 1 {
+		return fmt.Errorf("boulder/grpc/creds: peer had zero VerifiedChains")
+	}
+
+	/*
+	 * For each of the peer's verified chains we can look at the chain's leaf
+	 * certificate and check whether the subject common name is in the whitelist.
+	 * At least one chain must have a leaf certificate with a subject CN that
+	 * matches the whitelist
+	 *
+	 * Its important we process `VerifiedChains` instead of processing
+	 * `PeerCertificates` to ensure that we match the subject CN of the
+	 * leaf certificate that was verified in `conn.Handshake()`. To do otherwise
+	 * would allow an attacker to include a whitelisted certificate in
+	 * `PeerCertificates` that matched the whitelist but wasn't used in the chain
+	 * the server validated.
+	 */
+	var whitelisted bool
+	for _, chain := range chains {
+		leafSubjectCN := chain[0].Subject.CommonName
+		if _, ok := tc.whitelist[leafSubjectCN]; ok {
+			whitelisted = true
+		}
+	}
+
+	// If none of the peer's validated chains had a leaf certificate with a
+	// whitelisted CN then we have to reject the connection
+	if !whitelisted {
+		return fmt.Errorf(
+			"boulder/grpc/creds: peer's verified TLS chains did not include a leaf " +
+				"certificate with a whitelisted subject CN")
+	}
+
+	// Otherwise, the peer is whitelisted! Come on in!
+	return nil
+}
+
 // ServerHandshake performs the TLS handshake for a server <- client connection
```

Comment:
Could you update this comment to more closely match what's documented at https://godoc.org/google.golang.org/grpc/credentials#TransportCredentials? The current version isn't quite accurate, or at least is confusing.

Also, I'm particularly confused about one thing: Previously, `ServerHandshake` would have always returned an error. Which means it was never called, right? What has changed such that it will now start being called?